apollo-runtime/src/main/java/com/apollographql/apollo/internal/ApolloCallTracker.java

```diff
@@ -0,0 +1,174 @@
+package com.apollographql.apollo.internal;
+
+import com.apollographql.apollo.ApolloCall;
+import com.apollographql.apollo.ApolloPrefetch;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * ApolloCallTracker is responsible for keeping track of running {@link ApolloCall} & {@link ApolloPrefetch} objects.
+ */
+public class ApolloCallTracker {
+
+  private Runnable idleCallback;
+
+  private final Deque<ApolloCall> runningSyncCalls = new ArrayDeque<>();
+  private final Deque<RealApolloCall.AsyncCall> runningAsyncCalls = new ArrayDeque<>();
+  private final Deque<ApolloPrefetch> runningSyncPrefetches = new ArrayDeque<>();
+  private final Deque<RealApolloPrefetch.AsyncCall> runningAsyncPrefetches = new ArrayDeque<>();
+
+  public ApolloCallTracker() {
+  }
+
+  /**
+   * <p>Adds this {@link ApolloPrefetch} to the underlying data structure keeping track of the in progress synchronous
+   * prefetch objects.</p>
+   *
+   * <p><b>Note</b>: This method needs to be called right before a prefetch call is executed.</p>
+   */
+  synchronized void syncPrefetchInProgress(ApolloPrefetch apolloPrefetch) {
+    runningSyncPrefetches.add(apolloPrefetch);
+  }
+
+  /**
+   * <p>Removes this {@link ApolloPrefetch} from the underlying data structure keeping track of the in progress
+   * synchronous prefetch objects, if it is found, else throws an {@link AssertionError}.</p> If the removal operation
+   * is successful and no active running calls are found, then the registered {@link ApolloCallTracker#idleCallback} is
+   * invoked.
+   *
+   * <p><b>Note</b>: This method needs to be called right after a prefetch call is completed (whether successful or
+   * failed).</p>
+   */
+  void syncPrefetchFinished(ApolloPrefetch apolloPrefetch) {
+    Runnable idleCallback;
+    int runningCallsCount;
+    synchronized (this) {
+      if (!runningSyncPrefetches.remove(apolloPrefetch)) {
+        throw new AssertionError("Prefetch call wasn't in progress");
```

Comment:
Yeah, I feel that's an exceptional circumstance, because any part of code trying to remove a call object which was never added is a programmer's fault and we should stop then and there itself.